### ISSUE TITLE
ResourceExhaustedError when using TPU

### ISSUE THREAD

**Issue by gena** [googlecolab/colabtools#1470]
I have a few notebooks on Colab Pro which use TPU and worked perfectly a day ago, but now everything crashes

```
ResourceExhaustedError: 9 root error(s) found.
  (0) Resource exhausted: {{function_node __inference_train_function_453721}} Compilation failure: Ran out of memory in memory space hbm. Used 16.79G of 7.48G hbm. Exceeded hbm capacity by 9.31G.

```
Is there any changelog where I can see what did change in Colab, or this something to do with the TPU infrastructure? 

I can make it work by reducing the batch size but it has to be reduced like twice making models train at least 2x slower.

I've noticed that TensorFlow started to show strange warnings:

```
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/data/ops/multi_device_iterator_ops.py:601: get_next_as_optional (from tensorflow.python.data.ops.iterator_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.data.Iterator.get_next_as_optional()` instead.

WARNING:tensorflow:Callbacks method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0340s vs `on_train_batch_end` time: 0.4056s). Check your callbacks.

```

**Comment by craigcitro** [comment]
Is it possible to create a self-contained repro notebook?

**Comment by reyvaz** [comment]
I noticed they updated to tensorflow 2.3.0 yesterday and I also have been having issues with notebooks that ran perfectly on TPUs the day before while still on tensorflow 2.2.0. and that still run perfectly in Kaggle. 

I noticed also the new warnings, including the: 

`
WARNING:tensorflow:Callbacks method `on_train_batch_end` is slow compared to the batch time (batch time: 0.0115s vs `on_train_batch_end` time: 0.2912s). Check your callbacks
`

popping up after the first few batches. Which I was able to reproduce just by running this [colab tutorial notebook](https://colab.research.google.com/notebooks/tpu.ipynb); It appears during the 1st epoch. 

I have been having issues with my own notebooks "freezing" (i.e. not beginning to train, no OOM error either) with images larger than 380 (which trained perfectly before the tf 2.3.0 update). However, I have not been able to reproduce this with the flowers tutorial notebook above; it runs smoothly even with the size 512 flower dataset. 

I'll keep trying to create a self-contained repro with the flowers notebook. I'll get back here if I am able to, or if I find something new.

**Comment by gena** [comment]
It's an on-going Kaggle competition, I can share a notebook privately (mail or chat me on gennadiy.donchyts@gmail.com).

**Comment by gena** [comment]
It's an on-going Kaggle competition, I can share a notebook privately (chat or mail gennadiy.donchyts@gmail.com)

**Comment by linyuanthocr** [comment]
Please fix it as soon as possible. I am on the on-going Kaggle competition as well, the deadline is very close. Really need the colab to run experiments! Thanks!

**Comment by nguyenphuhien13** [comment]
I am having the same problem, really need a fix soon.

**Comment by craigcitro** [comment]
If you have repro steps, I'd encourage you to raise an issue directly in the TF repo: we're just bundling TF 2.3.

In particular, is this the same issue? https://github.com/tensorflow/tensorflow/issues/42043

**Comment by reyvaz** [comment]
I think I was able to reproduce the issue in this [repro notebook](https://colab.research.google.com/drive/1si9bdd0gvd4H0VktaVKYHg_QTUSoQoVy?usp=sharing) using the flowers dataset. 

@craigcitro I believe it is related to [tensorflow/tensorflow#42043](https://github.com/tensorflow/tensorflow/issues/42043) so I'm going to take a closer look there too. I have that same issue with pretty much all models that do not OOM like the one I'm sharing here.

**Comment by PuckWong** [comment]
I am having the same problem. Seems the tpu now only has 7.48G hbm.

**Comment by craigcitro** [comment]
For anyone looking to help get this fixed: making comments on the upstream issue is most helpful.

For anyone looking to use TF 2.2 with a TPU for now, this should get you unblocked:

```
!pip install tensorflow~=2.2.0 tensorflow_gcs_config~=2.2.0
import tensorflow as tf
import requests
import os
resp = requests.post("http://{}:8475/requestversion/{}".format(os.environ["COLAB_TPU_ADDR"].split(":")[0], tf.__version__))
if resp.status_code != 200:
  print("Failed to switch the TPU to TF {}".format(version))
```

**Comment by PuckWong** [comment]
> For anyone looking to help get this fixed: making comments on the upstream issue is most helpful.
> 
> For anyone looking to use TF 2.2 with a TPU for now, this should get you unblocked:
> 
> ```
> !pip install tensorflow~=2.2.0 tensorflow_gcs_config~=2.2.0
> import tensorflow as tf
> import requests
> import os
> resp = requests.post("http://{}:8475/requestversion/{}".format(os.environ["COLAB_TPU_ADDR"].split(":")[0], tf.__version__))
> if resp.status_code != 200:
>   print("Failed to switch the TPU to TF {}".format(version))
> ```

switch to TF 2.2 works.

**Comment by linyuanthocr** [comment]
It works, thanks a lot!

**Comment by gena** [comment]
Thanks, @craigcitro!! This will work for now before it's fixed in 2.3+

**Comment by gena** [comment]
It broke again, and the above workaround does not work anymore :(

**Comment by gena** [comment]
It crashes again starting yesterday, and the workaround does not work anymore - @craigcitro, help!!! 😭 

Here is a gist trying to apply the TF2.2 workaround, using the notebook posted by @reyvaz: 
https://colab.research.google.com/gist/gena/ccb3e4f43dd9980b7e275c5a2b145075/tf_tpu_issue_submission.ipynb

**Comment by chucksylar** [comment]
The workaround stop working.... plz help....

**Comment by graf10a** [comment]
I am facing the same issue (see reyvaz's notebook above). Please let us use TF 2.2 until this TPU bug is fixed in TF 2.3.

**Comment by gena** [comment]
@graf10a, just to clarify, the @reyvaz's notebook tries to use TF2.3 (current default) and fails, the notebook in my last comment tries to use the workaround posted by @craigcitro, and currently fails well. We need at least some working Colab, the current version is basically can't be used for models which worked a few days ago.

**Comment by graf10a** [comment]
@gena Sorry, I just ran the notebook at the link you provided and got the same error. This is what I meant. Actually, I made some changes in this notebook and managed to make it work: 
https://colab.research.google.com/drive/1vd0_3KIrxy8Kw1OSfmJvw8eoZSrZ1bXh?usp=sharing

Also, I ran one of my own old notebooks with the TF 2.2 workaround and it seems to be working just fine. So, I am not sure what is the issue here (but I might be missing something).

**Comment by gena** [comment]
@graf10a, thanks for letting know this, interesting, I'm actually able to train B0 with this workaround, but without jumping into larger resolution and batch sizes. Maybe my old notebooks are somehow stuck with a TPU not switching context ... but it returns 200, strange.

**Comment by graf10a** [comment]
@gena Did you notice that I made some changes in the TPU initialization part of the notebook? I don't think this part was working properly in your version of the notebook. Also, I have changed the import of EfficentNet model. Maybe this was what made the difference?

**Comment by gena** [comment]
Yes, this is also how I initialized it. I guess we should then stay with small batch sizes and I'm also using the same EFN, but somehow getting OOM on 256x256 batch size=16, strange, and sometimes it runs.

**Comment by graf10a** [comment]
@gena Great! I am glad it is working for you now!

**Comment by colaboratory-team** [comment]
Duplicate of https://github.com/tensorflow/tensorflow/issues/42043
Please do follow up on the TensorFlow issue. We'll need to pursue a fix for this in TensorFlow rather than Colab.
